### PR TITLE
Recommend "when" keyword after case-pattern-switch-label

### PR DESCRIPTION
**Pull request by jcouv**
### Customer scenario
Type a switch/case with a pattern, such as `case DeclarationExpressionSyntax declExpr $$`. At this point, `when` should be offered as a completion.

### Bugs this fixes
Fixes https://github.com/dotnet/roslyn/issues/24113

### Workarounds, if any
Type "when" manually.

### Performance impact
Low. The change only impacts the when keyword recommender and only adds a couple of kind checks on the syntax.

### Is this a regression from a previous update?
No

### How was the bug found?
Reported by me. Discovered by typing such code.

### CODE REVIEW

**Comment by jcouv** [comment]
@dotnet/roslyn-ide for review. Thanks
FYI @gafter

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs [hunk 1]
```diff
@@ -14,7 +14,31 @@ public WhenKeywordRecommender()
...
+            }
+
+            // case int i $$
+            if (IsIdentifierInCasePattern(context.LeftToken.Parent))
+            {
+                return true;
+            }
+
+            // case int i w$$
+            if (IsIdentifierInCasePattern(context.LeftToken.GetPreviousToken().Parent))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs]: this shouldn't be needed.  see other recommenders for how they handle both the unwritten and partially written case.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs]: I think you can jsut use "context.TargetToken" for this.

File: src/EditorFeatures/CSharpTest2/Recommendations/RecommenderTests.cs [hunk 2]
```diff
@@ -82,8 +82,8 @@ private Task CheckResultAsync(string text, int position, bool absent, CSharpPars
                 }
                 else
                 {
-                    var result = (await RecommendKeywordsAsync(position, context)).Single();
-                    Assert.NotNull(result);
+                    var result = (await RecommendKeywordsAsync(position, context))?.Single();
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest2/Recommendations/RecommenderTests.cs]: ❓ Why would you need `?.` here?

:bulb: Consider the following:

```csharp
var result = await RecommendKeywordsAsync(position, context);
Assert.Single(result);
```

File: src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs [hunk 3]
```diff
@@ -63,5 +63,13 @@ public class WhenKeywordRecommenderTests : KeywordRecommenderTests
             await VerifyAbsenceAsync(AddInsideMethod(
 @"try {} catch (Exception e) when (true) $$"));
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs]: :bulb: Consider adding tests corresponding to each of the edge cases added in `IsValidContext`. Ideally one or more of the tests should fail if any of the `if` statements is removed.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs [hunk 4]
```diff
@@ -14,7 +14,31 @@ public WhenKeywordRecommender()
...
+                return true;
+            }
+
+            return false;
+        }
+
+        private static bool IsIdentifierInCasePattern(SyntaxNode node)
+        {
+            return node.IsKind(SyntaxKind.SingleVariableDesignation)
+                && node.Parent.IsKind(SyntaxKind.DeclarationPattern)
```
Comment (sharwell) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs]: :bulb: `node.IsParentKind(SyntaxKind.DeclarationPattern)`

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs [hunk 5]
```diff
@@ -14,7 +14,31 @@ public WhenKeywordRecommender()
...
+                && node.Parent.Parent.IsKind(SyntaxKind.CasePatternSwitchLabel);
```
Comment (sharwell) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs]: 💡 `node.Parent.IsParentKind(SyntaxKind.CasePatternSwitchLabel)`

File: src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs (hunk 3 above)
Comment (jcouv) [reply src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs]: This test already covers all those cases (it includes multiple variants, typing the keyword one character at a time) :-)

File: src/EditorFeatures/CSharpTest2/Recommendations/RecommenderTests.cs (hunk 2 above)
Comment (jcouv) [reply src/EditorFeatures/CSharpTest2/Recommendations/RecommenderTests.cs]: Good catch. The `?.` was unnecessary. Removed it

Having a message in the assert is what I'm after, so I'm leaving it. That improves the experience working in IDE codebase. Below is the generic message assert message the old code would produce (not very informative):

```
Test 'Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Recommendations.WhenKeywordRecommenderTests.TestInCasePattern' failed:
	System.ArgumentNullException : Value cannot be null.
Parameter name: source
	at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source)
	at Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Recommendations.RecommenderTests.<CheckResultAsync>d__5.MoveNext()
...
```

**Comment by jcouv** [comment]
@sharwell Could you take another look when you get a chance? Addressed your feedback. Thanks

**Comment by jcouv** [comment]
@Pilchie Please approve for ask-mode for 15.6. Thanks

**Pull request merged**